Implement GGML_CPU_ALL_VARIANTS for ARM

ckastner opened the pull request:
This supersedes #14049 which also has more context.

There are two notable design decisions, better explained in the respective commit messages:
1. The use of GGML_INTERNAL_<FEAT> within cmake, and the GGML_USE_<FEAT> this activates in code
2. The odd backend naming as a consequence of features being optional

I tested this on a 4-vcpu Graviton4 which is armv9.0-a. Test command was simply `llama-bench -m ggml-model-q4_0.gguf`, as I just needed something simple to show that loading worked correctly and that no regressions were introduced.

First, the results with `GGML_NATIVE=ON`:

```shell
| model                          |       size |     params | backend    | threads |            test |                  t/s |
| ------------------------------ | ---------: | ---------: | ---------- | ------: | --------------: | -------------------: |
| llama 1B Q4_0                  | 606.54 MiB |     1.10 B | CPU        |       4 |           pp512 |         38.57 ± 0.01 |
| llama 1B Q4_0                  | 606.54 MiB |     1.10 B | CPU        |       4 |           tg128 |          6.77 ± 0.00 |
```

Then, the results of `GGML_NATIVE=OFF GGML_BACKEND_DL=ON GGML_CPU_ALL_VARIANTS=ON` (some debug messages omitted):

```shell
ggml_backend_load_best: /home/christian/GitHub/llama.cpp/build/bin/libggml-cpu-armv9.2_2.so score: 0
ggml_backend_load_best: /home/christian/GitHub/llama.cpp/build/bin/libggml-cpu-armv8.2_3.so score: 15
ggml_backend_load_best: /home/christian/GitHub/llama.cpp/build/bin/libggml-cpu-armv8.2_1.so score: 3
ggml_backend_load_best: /home/christian/GitHub/llama.cpp/build/bin/libggml-cpu-armv8.2_2.so score: 7
ggml_backend_load_best: /home/christian/GitHub/llama.cpp/build/bin/libggml-cpu-armv8.6_2.so score: 63
ggml_backend_load_best: /home/christian/GitHub/llama.cpp/build/bin/libggml-cpu-armv8.0_1.so score: 1
ggml_backend_load_best: /home/christian/GitHub/llama.cpp/build/bin/libggml-cpu-armv9.2_1.so score: 0
ggml_backend_load_best: /home/christian/GitHub/llama.cpp/build/bin/libggml-cpu-armv8.6_1.so score: 31
load_backend: loaded CPU backend from /home/christian/GitHub/llama.cpp/build/bin/libggml-cpu-armv8.6_2.so

| model                          |       size |     params | backend    | threads |            test |                  t/s |
| ------------------------------ | ---------: | ---------: | ---------- | ------: | --------------: | -------------------: |
| llama 1B Q4_0                  | 606.54 MiB |     1.10 B | CPU        |       4 |           pp512 |         38.58 ± 0.01 |
| llama 1B Q4_0                  | 606.54 MiB |     1.10 B | CPU        |       4 |           tg128 |          6.77 ± 0.00 |
```
The scoring for each backend was calculated correctly.

The armv9.2 backends are for SME which the Graviton4 doesn't have, hence they scored 0. So the armv8.6_2 backend (`+dotprod+fp16+sve+i8mm`) got picked.

Incidentally, this showcases one problem that I left for future work. When choosing the MCPU to target, I chose the first version that supported a particular instruction, which eg: for `i8mm` was armv8.6-a. However, the test above was run on armv9.0-a, so a build with same features but targeting armv9.0-a might have performed even better. The solution would be to include the runtime arch in the scoring, but the above implements the necessary base case and I'll look to this improvement later.

ckastner committed:
ggml-cpu: Add ARM feature detection and scoring

This is analogous to cpu-feats-x86.cpp. However, to detect compile-time
activation of features, we rely on GGML_USE_<FEAT> which need to be set
in cmake, instead of GGML_<FEAT> that users would set for x86.

This is because on ARM, users specify features with GGML_CPU_ARM_ARCH,
rather than with individual flags.

ckastner committed:
ggml-cpu: Implement GGML_CPU_ALL_VARIANTS for ARM

Like x86, however to pass around arch flags within cmake, we use
GGML_INTERNAL_<FEAT> as we don't have GGML_<FEAT>.

Some features are optional, so we may need to build multiple backends
per arch version (armv8.2_1, armv8.2_2, ...), and let the scoring
function sort out which one can be used.

ckastner committed:
ggml-cpu: Factor out feature detection build from x86

slaren commented:
We are going to need a different set of variants for each platform. For example I would expect the variants used for Apple to be each of the M1-M4 chips (or at least the ones that have different CPU features). The Android variants are also likely to be different than the Linux variants. Windows at the moment probably only needs one variant. It's not necessary to support every platform from the first moment, but this list of variants probably should only be used for Linux.

ckastner commented:
> We are going to need a different set of variants for each platform. [...] It's not necessary to support every platform from the first moment, but this list of variants probably should only be used for Linux.

I agree, I'll add that later today.

> Generally I think it is easier to build the list of variants if we know exactly the list of chips we are targeting.

Same, though I saw this as a topic for future iterations once the basic mechanism is in place. Might also be an opportunity for the respective manufacturers to chip in (pun intended).

ckastner committed:
ggml-cpu: Limit ARM GGML_CPU_ALL_VARIANTS to Linux for now

The other platforms will need their own specific variants.

This also fixes the bug that the the variant-building branch was always
being executed as the else-branch of GGML_NATIVE=OFF. The branch is
moved to an elseif-branch which restores the previous behavior.

chaxu01 commented:
Benchmarked on Graviton3:

W/O GGML_CPU_ALL_VARIANTS
```
$TOOLCHAIN_SYSROOT/lib/ld-linux-aarch64.so.1 --library-path $TOOLCHAIN_SYSROOT/lib64:$TOOLCHAIN_SYSROOT/usr/lib64:/tmp/chaxu01 ./llama-bench -m llama-2-7b-chat.Q4_0.gguf -ngl 0 -t 1,2,4,8,16
load_backend: loaded CPU backend from /tmp/chaxu01/libggml-cpu-armv8.6_1.so
| model                          |       size |     params | backend    | threads |            test |                  t/s |
| ------------------------------ | ---------: | ---------: | ---------- | ------: | --------------: | -------------------: |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CPU        |       1 |           pp512 |         15.17 ± 0.01 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CPU        |       1 |           tg128 |          5.73 ± 0.01 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CPU        |       2 |           pp512 |         30.26 ± 0.04 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CPU        |       2 |           tg128 |         10.46 ± 0.01 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CPU        |       4 |           pp512 |         60.23 ± 0.04 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CPU        |       4 |           tg128 |         18.25 ± 0.00 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CPU        |       8 |           pp512 |        101.29 ± 0.05 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CPU        |       8 |           tg128 |         30.75 ± 0.03 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CPU        |      16 |           pp512 |        178.37 ± 0.46 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CPU        |      16 |           tg128 |         40.93 ± 0.21 |

build: f5a8b9ad (5619)
```

```
W/ GGML_CPU_ALL_VARIANTS
$TOOLCHAIN_SYSROOT/lib/ld-linux-aarch64.so.1 --library-path $TOOLCHAIN_SYSROOT/lib64:$TOOLCHAIN_SYSROOT/usr/lib64:/tmp/chaxu01 ./llama-bench -m llama-2-7b-chat.Q4_0.gguf -ngl 0 -t 1,2,4,8,16
load_backend: loaded CPU backend from /tmp/chaxu01/libggml-cpu-armv8.6_1.so
| model                          |       size |     params | backend    | threads |            test |                  t/s |
| ------------------------------ | ---------: | ---------: | ---------- | ------: | --------------: | -------------------: |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CPU        |       1 |           pp512 |         15.15 ± 0.01 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CPU        |       1 |           tg128 |          5.72 ± 0.01 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CPU        |       2 |           pp512 |         30.17 ± 0.01 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CPU        |       2 |           tg128 |         10.44 ± 0.01 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CPU        |       4 |           pp512 |         60.24 ± 0.03 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CPU        |       4 |           tg128 |         18.18 ± 0.01 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CPU        |       8 |           pp512 |        101.18 ± 0.09 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CPU        |       8 |           tg128 |         30.50 ± 0.09 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CPU        |      16 |           pp512 |       133.59 ± 34.84 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | CPU        |      16 |           tg128 |        35.80 ± 10.28 |

build: f5a8b9ad (5619)
```
Not sure why the number for threads 16 were off for W/ GGML_CPU_ALL_VARIANTS, otherwise it looks good. Looking forward adding the support for Apple and Android.

ckastner commented:
Thanks for the check!

Are you sure the first example is with `GGML_CPU_ALL_VARIANTS=OFF`? It loaded `libggml-cpu-armv8.6_1.so`, a name which should only appear with =ON. The =OFF version should just be called `libggml-cpu.so`.

I tested it with `GGML_BACKEND_DL=ON` and then `GGML_NATIVE=ON GGML_ALL_CPU_VARIANTS=OFF`, and vice versa.

slaren reviewed (approved):
I don't have a linux arm machine to test this, but the code looks good to me. The only thing I don't like very much is how the base arm arch is set automatically depending on which features are added, it is not obvious to me that this will always be correct. But it can changed if necessary when adding support for other platforms.

ckastner commented:
> The only thing I don't like very much is how the base arm arch is set automatically depending on which features are added

Yeah this definitely has room for future improvement. Regarding correctness, I assumed that if some instruction was only added in arch version X, then it is safe to bump the base arch to X and possibly get other improvements. @chaxu01, your thoughts on that?

ckastner commented:
> In my case, I check for compiler support for arch X for each feature and add the supported variant.

A fair point. It's even a bit more complicated I think, for example the support check is the same that is done for GGML_NATIVE=ON so that could be factored out.  Like `FindSIMD` for x86 does.

Then there's a choice where to check and skip the build gif not supported: before or in the _variant function). Before is cleaner I think, but it also feels a bit like a layer violation given the current setup.

@slaren this new question seems nuanced enough to warrant its own PR, is that ok or would you like that addressed here?
